### ISSUE TITLE
Autofac.Core.Registration.ComponentNotRegisteredException

### ISSUE THREAD

**Issue by Codefans-fan** [nopSolutions/nopCommerce#3264]
nopCommerce version: 4.1

Start the app from source code. In debug out put log I get these exception

> 'dotnet.exe' (CoreCLR: clrhost): Loaded 'j2o3nqrb.rzh'. Cannot find or open the PDB file.

> 'dotnet.exe' (CoreCLR: clrhost): Loaded 'j2o3nqrb.rzh'. Cannot find or open the PDB file.

> Exception thrown: 'Autofac.Core.Registration.ComponentNotRegisteredException' in Autofac.dll

> Exception thrown: 'Autofac.Core.Registration.ComponentNotRegisteredException' in Autofac.dll

what the reason to cause this exception ?

**Comment by AndreiMaz** [comment]
please provide a list of steps to reproduce the issue

**Comment by Codefans-fan** [comment]
1. download 4.10 source code.
2. Run the source code without change any thing.
3.  These exception would show in console

If  I uninstall all plugins. I would not get the exception.

**Comment by RomanovM** [comment]
I just tried to reproduce the problem and launched a clean solution. It works like a charm without any errors.

**Comment by Codefans-fan** [comment]
@RomanovM have you install some official plugins ?    clean solution would not get this issue.
I install some payment plugin I get this issue.  It seem if plugin contain services or DependencyRegistrar have this problem. would you try to install some plugin and check again.  Thank you

**Comment by brakim** [comment]
I'm experiencing same problem.

**Comment by alecrt** [comment]
Same problem here

**Comment by alecrt** [comment]
More details: 

error happens to me if I have at least one plugin (official or not) in Nop.Core.Plugins.Instance method:

```/// <summary>
        /// Get the instance of the plugin
        /// </summary>
        /// <typeparam name="T">Type of the plugin</typeparam>
        /// <returns>Plugin instance</returns>
        public virtual T Instance<T>() where T : class, IPlugin
        {
            object instance = null;
            try
            {
                instance = EngineContext.Current.Resolve(PluginType);
            }
            catch
            {
                //try resolve <--- ERROR CATCHED HERE
            }

            if (instance == null)
            {
                //not resolved
                instance = EngineContext.Current.ResolveUnregistered(PluginType);
            }

            var typedInstance = instance as T;
            if (typedInstance != null)
                typedInstance.PluginDescriptor = this;

            return typedInstance;
        }` 

Example of exception thrown: 
'Autofac.Core.Registration.ComponentNotRegisteredException' in Autofac.dll ("The requested service 'Nop.Plugin.Tax.FixedOrByCountryStateZip.FixedOrByCountryStateZipTaxProvider' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency.") Exception thrown: 'Autofac.Core.Registration.ComponentNotRegisteredException' in Autofac.dll ("The requested service 'Nop.Plugin.Tax.FixedOrByCountryStateZip.FixedOrByCountryStateZipTaxProvider' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency."

**Comment by glararan** [comment]
Hi,

I have same issue, downloaded source code via website. Compiled, installed new database, launched and then exceptions came out.

`Exception thrown: 'System.ArgumentException' in System.Private.CoreLib.dll ("The 'AppId' option must be provided.")`
`The 'AppId' option must be provided.`

`Exception thrown: 'Autofac.Core.Registration.ComponentNotRegisteredException' in Autofac.dll ("The requested service 'Nop.Plugin.Widgets.GoogleAnalytics.GoogleAnalyticsPlugin' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency.")`
`The requested service 'Nop.Plugin.Widgets.GoogleAnalytics.GoogleAnalyticsPlugin' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency.`

`Exception thrown: 'Autofac.Core.Registration.ComponentNotRegisteredException' in Autofac.dll ("The requested service 'Nop.Plugin.Widgets.NivoSlider.NivoSliderPlugin' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency.")`
`The requested service 'Nop.Plugin.Widgets.NivoSlider.NivoSliderPlugin' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency.`

**Comment by Codefans-fan** [comment]
Debug into NopEngine.cs .
I find the method try to  get service  using plugin type name. 
add these code seems would fix it
` public object Resolve(Type type)
        {
            if (type.IsSubclassOf(typeof(BasePlugin))) {
                return null;
            }

            return GetServiceProvider().GetRequiredService(type);
        }`

**Comment by alecrt** [comment]
I dunno if it’s the right place to fix it. I’d fix it on Nop.Core.Plugins.PluginDescriptor Instance method, as plugins are not registered in the service container anywhere. 

But only @AndreiMaz knows this for sure. I only hope he’ll get on this issue as is marked as closed.

**Comment by softkey87** [comment]
On my side, I solved this using GetService(type) instead of GetRequiredService(type), which return null instead of throwing an error. I don't know if it's OK to do so, but I'm running this since two week without problems.

**Comment by skoshelev** [comment]
@Codefans-fan, alecrt and softkey87 Thank you for your help, we analyzed the situation and the debug log, and in the end we implemented a modular solution from the pieces you offered

Closed #3264
